client/consensus/grandpa/authorities.go

```diff
@@ -160,51 +231,414 @@ func (authSet *AuthoritySet) addForcedChange(pending PendingChange, isDescendent
+	if err != nil {
+		return nil, fmt.Errorf("%w: %v", errForkTree, err)
+	}
+	return applied, nil
 }
 
-// InnerLocked
-// Returns access to the [`AuthoritySet`] and locks it.
+// PendingChange A pending change to the authority set.
 //
-// For more information see [`SharedDataLocked`].
-// TODO update with shared logic
-func (sas *SharedAuthoritySet) InnerLocked() AuthoritySet {
-	return sas.authoritySet
+// This will be applied when the announcing block is at some depth within
+// the finalized or unfinalized chain.
+type PendingChange struct {
+	// The new authorities and weights to apply.
+	nextAuthorities AuthorityList
+	// How deep in the chain the announcing block must be
+	// before the change is applied.
+	delay uint
+	// The announcing block's height.
+	canonHeight uint
+	// The announcing block's hash.
+	canonHash common.Hash
+	// The delay kind.
+	delayKind DelayKind
+}
+
+// EffectiveNumber Returns the effective number this change will be applied at.
+func (pc *PendingChange) EffectiveNumber() uint {
+	return pc.canonHeight + pc.delay
+}
+
+// AuthoritySetChanges Tracks historical authority set changes. We store the block numbers for the last block
+// of each authority set, once they have been finalized. These blocks are guaranteed to
+// have a justification unless they were triggered by a forced change.
+type AuthoritySetChanges []authorityChange
+
+// append an authorityChange to AuthoritySetChanges
+func (authSetChanges *AuthoritySetChanges) append(setId uint64, blockNumber uint) {
+	*authSetChanges = append(*authSetChanges, authorityChange{
+		setId:       setId,
+		blockNumber: blockNumber,
+	})
+}
+
+// Three states that can be returned: Latest, Set (tuple), Unknown
+// Latest => bool
+// Set => &AuthorityChange
+// Unknown => nil
+// TODO for reviewers, this can be a VDT but I'm not sure its needed
+func (authSetChanges *AuthoritySetChanges) getSetId(blockNumber uint) (latest bool, set *authorityChange, err error) {
+	if authSetChanges == nil {
+		return false, nil, fmt.Errorf("getSetId: authSetChanges is nil")
+	}
+	authSet := *authSetChanges
+	last := authSet[len(authSet)-1]
+	if last.blockNumber < blockNumber {
+		return true, nil, nil // Latest case
+	}
+
+	idx, _ := searchSetChanges(blockNumber, authSet)
+	if idx < len(authSet) {
+		authChange := authSet[idx]
+
+		// if this is the first index but not the first set id then we are missing data.
+		if idx == 0 && authChange.setId != 0 {
+			return false, nil, nil // Unknown case
+		}
+
+		return false, &authChange, nil // Set case
+	}
+
+	return false, nil, nil // Unknown case
+}
+
+func (authSetChanges *AuthoritySetChanges) insert(blockNumber uint) {
+	var idx int
+	if authSetChanges == nil {
+		idx = 0
+	} else {
+		idx, _ = searchSetChanges(blockNumber, *authSetChanges)
+	}
+
+	set := *authSetChanges
```

Comment:
if `authSetChanges == nil` and you deref it will panic and there is no way to initialize it from inside this metho since the initialization will not take effect outside ref: [nil-receiver-in-golang](https://medium.com/@reetas/nil-receiver-in-golang-9d61ed8fd230) so I would suggest to `if autSetChanges == nil` then `panic("authority set changes must be initialized")`

Reply:
removed the set index to 0 and replaced it with a panic, good call